### PR TITLE
support in-place update operation for hnsw

### PR DESCRIPTION
**Pull request by ShawnShawnYou**
* support in-place update operation for hnsw #225

### CODE REVIEW

File: tests/test_index_old.cpp [hunk 1]
```diff
@@ -637,6 +637,90 @@ TEST_CASE("remove vectors from the index", "[ft][index]") {
     }
 }
 
+TEST_CASE("update vectors from the index", "[ft][index]") {
```
Comment (inabao) [inline tests/test_index_old.cpp]: add thread-safe testing for hnsw

Comment (ShawnShawnYou) [reply tests/test_index_old.cpp]: done

**Review by inabao — APPROVED** [review]
LGTM

File: tests/test_index_old.cpp [hunk 2]
```diff
@@ -637,6 +637,122 @@ TEST_CASE("remove vectors from the index", "[ft][index]") {
     }
 }
 
+TEST_CASE("update id and vectors from the index", "[ft][index]") {
```
Comment (LHT129) [inline tests/test_index_old.cpp]: use new index test template for new interface functest

File: include/vsag/index.h [hunk 3]
```diff
@@ -88,6 +88,31 @@ class Index {
...
+     * @param old_id indicates the old id of a base point in index
+     * @param new_id is the updated new id of the base point
+     * @return result indicates whether the update operation is successful.
+     */
+    virtual tl::expected<bool, Error>
+    UpdateId(int64_t old_id, int64_t new_id) {
+        throw std::runtime_error("Index not support update id");
+    }
+
+    /**
```
Comment (LHT129) [inline include/vsag/index.h:103]: add new feature on feature.h?

File: tests/test_multi_thread.cpp [hunk 4]
```diff
@@ -293,6 +295,31 @@ TEST_CASE("multi-threading read-write test", "[ft][hnsw]") {
             return add_res.value().size();
         }));
 
+        // update id
+        update_id_results.push_back(
```
Comment (LHT129) [inline tests/test_multi_thread.cpp:240]: use new multithread test for update test

File: include/vsag/index.h [hunk 5]
```diff
@@ -88,6 +88,31 @@ class Index {
         throw std::runtime_error("Index not support delete vector");
     }
 
+    /**
+     * Update the id of a base point from the index
+     *
+     * @param old_id indicates the old id of a base point in index
+     * @param new_id is the updated new id of the base point
+     * @return result indicates whether the update operation is successful.
+     */
```
Comment (LHT129) [inline include/vsag/index.h:97]: what is difference between `return false` and `error`

File: src/algorithm/hnswlib/hnswalg.cpp [hunk 6]
```diff
@@ -1136,6 +1136,43 @@ HierarchicalNSW::dealNoInEdge(InnerIdType id, int level, int m_curmax, int skip_
...
+    std::unique_lock<std::mutex> lock(global_);
+    auto iter = label_lookup_.find(label);
+    if (iter == label_lookup_.end()) {
+        throw std::runtime_error("no label in HNSW");
+    } else {
+        internal_id = iter->second;
+
+        // reset data
+        std::shared_ptr<float[]> normalize_data;
+        normalizeVector(data_point, normalize_data);
```
Comment (LHT129) [inline src/algorithm/hnswlib/hnswalg.cpp:1150]: need allocator?

File: src/algorithm/hnswlib/hnswalg.cpp [hunk 7]
```diff
@@ -1136,6 +1136,43 @@ HierarchicalNSW::dealNoInEdge(InnerIdType id, int level, int m_curmax, int skip_
...
+        memcpy(getDataByInternalId(internal_id), data_point, data_size_);
```
Comment (LHT129) [inline src/algorithm/hnswlib/hnswalg.cpp:1151]: why use data_point rathe than normalize_data

**Review by LHT129 — CHANGES_REQUESTED** [review]

File: include/vsag/index.h [hunk 8]
```diff
@@ -88,6 +88,31 @@ class Index {
...
+    virtual tl::expected<bool, Error>
+    UpdateId(int64_t old_id, int64_t new_id) {
```
Comment (jiaweizone) [inline include/vsag/index.h:99]: Why need update id ?

File: src/algorithm/hnswlib/hnswalg.cpp [hunk 9]
```diff
@@ -1136,6 +1136,43 @@ HierarchicalNSW::dealNoInEdge(InnerIdType id, int level, int m_curmax, int skip_
...
 
+void
+HierarchicalNSW::updateVector(LabelType label, const void* data_point) {
+    InnerIdType internal_id = 0;
+    std::unique_lock<std::mutex> lock(global_);
+    auto iter = label_lookup_.find(label);
+    if (iter == label_lookup_.end()) {
+        throw std::runtime_error("no label in HNSW");
+    } else {
+        internal_id = iter->second;
```
Comment (jiaweizone) [inline src/algorithm/hnswlib/hnswalg.cpp]: merge internal_id define to here
`InnerIdType internal_id = iter->second;`

File: src/algorithm/hnswlib/hnswalg.cpp (hunk 6 above)
Comment (jiaweizone) [inline src/algorithm/hnswlib/hnswalg.cpp:1150]: minor: Why `normalizeVector` need change input param `data_point` ?
This update behavior means the `data_point` pointer memory life cycle changed

File: src/algorithm/hnswlib/hnswalg.cpp [hunk 10]
```diff
@@ -1136,6 +1136,43 @@ HierarchicalNSW::dealNoInEdge(InnerIdType id, int level, int m_curmax, int skip_
     }
 }
 
+void
+HierarchicalNSW::updateVector(LabelType label, const void* data_point) {
+    InnerIdType internal_id = 0;
+    std::unique_lock<std::mutex> lock(global_);
+    auto iter = label_lookup_.find(label);
+    if (iter == label_lookup_.end()) {
+        throw std::runtime_error("no label in HNSW");
```
Comment (jiaweizone) [inline src/algorithm/hnswlib/hnswalg.cpp]: minor: add the label id content to error message ?

File: src/index/hnsw.cpp [hunk 11]
```diff
@@ -616,6 +616,54 @@ HNSW::GetStats() const {
...
+
+tl::expected<bool, Error>
+HNSW::update_vector(int64_t id, const DatasetPtr& new_base, bool need_fine_tune) {
+    if (use_static_) {
+        LOG_ERROR_AND_RETURNS(ErrorType::UNSUPPORTED_INDEX_OPERATION,
+                              "static hnsw does not support update");
+    }
+
+    try {
+        std::unique_lock lock(rw_mutex_);
```
Comment (jiaweizone) [inline src/index/hnsw.cpp]: the lock is not necessary ？

File: src/index/hnsw.cpp [hunk 12]
```diff
@@ -616,6 +616,54 @@ HNSW::GetStats() const {
...
+        spdlog::warn(
+            "update error for replace old_id {} to new_id {}: {}", old_id, new_id, e.what());
+        return false;
+    }
+
+    return true;
+}
+
+tl::expected<bool, Error>
+HNSW::update_vector(int64_t id, const DatasetPtr& new_base, bool need_fine_tune) {
```
Comment (jiaweizone) [inline src/index/hnsw.cpp:640]: `need_fine_tune` not use ?

Comment (ShawnShawnYou) [reply src/index/hnsw.cpp:640]: Remain it in the interface for further optimization ? (e.g., when distance is greater than specific threshold, we need to fine tune the graph structure)

File: src/index/hnsw.cpp (hunk 11 above)
Comment (ShawnShawnYou) [reply src/index/hnsw.cpp]: done

File: src/algorithm/hnswlib/hnswalg.cpp (hunk 10 above)
Comment (ShawnShawnYou) [reply src/algorithm/hnswlib/hnswalg.cpp]: done

File: src/algorithm/hnswlib/hnswalg.cpp (hunk 9 above)
Comment (ShawnShawnYou) [reply src/algorithm/hnswlib/hnswalg.cpp]: done

File: include/vsag/index.h (hunk 8 above)
Comment (ShawnShawnYou) [reply include/vsag/index.h:99]: For some cases, the users may only want to update the id or vector of a base point.  Then we need to provide two functions to provide flexibility.

File: include/vsag/index.h (hunk 5 above)
Comment (ShawnShawnYou) [reply include/vsag/index.h:97]: `error` means un-supported index type or config conflicts. `false` means update failed.

File: include/vsag/index.h (hunk 3 above)
Comment (ShawnShawnYou) [reply include/vsag/index.h:103]: done

File: tests/test_index_old.cpp (hunk 2 above)
Comment (ShawnShawnYou) [reply tests/test_index_old.cpp]: done

File: tests/test_multi_thread.cpp (hunk 4 above)
Comment (ShawnShawnYou) [reply tests/test_multi_thread.cpp:240]: remains for next PR?

File: src/algorithm/hnswlib/hnswalg.cpp (hunk 6 above)
Comment (ShawnShawnYou) [reply src/algorithm/hnswlib/hnswalg.cpp:1150]: this usage is align with `addPoint`, can be fixed in other PR

File: src/algorithm/hnswlib/hnswalg.cpp (hunk 7 above)
Comment (ShawnShawnYou) [reply src/algorithm/hnswlib/hnswalg.cpp:1151]: this usage is align with `addPoint`, can be fixed in other PR

File: src/algorithm/hnswlib/hnswalg.cpp (hunk 6 above)
Comment (ShawnShawnYou) [reply src/algorithm/hnswlib/hnswalg.cpp:1150]: this usage is align with `addPoint`, can be fixed in other PR?

Comment (jiaweizone) [reply src/algorithm/hnswlib/hnswalg.cpp:1150]: Ok, fix in next PR

File: src/index/hnsw.cpp (hunk 12 above)
Comment (jiaweizone) [reply src/index/hnsw.cpp:640]: Add TODO

Comment (ShawnShawnYou) [reply src/index/hnsw.cpp:640]: done

**Review by jiaweizone — APPROVED** [review]
LGTM

**Review by LHT129 — APPROVED** [review]
LGTM

**Pull request merged**